contracts/Aavegotchi/ForgeDiamond/facets/ForgeTokenFacet.sol

```diff
@@ -0,0 +1,302 @@
+
+    // @notice return users balance of all tokens
+    function balanceOfOwner(address account) public view returns (ItemBalancesIO[] memory output_) {
+        uint256 count = s.ownerItems[account].length;
+        output_ = new ItemBalancesIO[](count);
+        for (uint256 i; i < count; i++) {
+            uint256 tokenId = s.ownerItems[account][i];
+            output_[i].balance = s.ownerItemBalances[account][tokenId];
+            output_[i].tokenId = tokenId;
+        }
+    }
+
+    /**
+     * @dev See {IERC1155-setApprovalForAll}.
+     */
+    function setApprovalForAll(address operator, bool approved) public {
+        _setApprovalForAll(LibMeta.msgSender(), operator, approved);
+    }
+
+    /**
+     * @dev See {IERC1155-isApprovedForAll}.
+     */
+    function isApprovedForAll(address account, address operator) public view returns (bool) {
+        return s._operatorApprovals[account][operator];
+    }
+
+    /**
+     * @dev See {IERC1155-safeTransferFrom}.
+     */
+    function safeTransferFrom(
+        address from,
+        address to,
+        uint256 id,
+        uint256 amount,
+        bytes memory data
+    ) public {
+        require(
+            from == LibMeta.msgSender() || isApprovedForAll(from, LibMeta.msgSender()),
+            "ForgeTokenFacet: caller is not token owner or approved"
+        );
+        _safeTransferFrom(from, to, id, amount, data);
+    }
+
+    /**
+     * @dev See {IERC1155-safeBatchTransferFrom}.
+     */
+    function safeBatchTransferFrom(
+        address from,
+        address to,
+        uint256[] memory ids,
+        uint256[] memory amounts,
+        bytes memory data
+    ) public {
+        require(
+            from == LibMeta.msgSender() || isApprovedForAll(from, LibMeta.msgSender()),
+            "ForgeTokenFacet: caller is not token owner or approved"
+        );
+        _safeBatchTransferFrom(from, to, ids, amounts, data);
+    }
+
+    /**
+     * @dev Transfers `amount` tokens of token type `id` from `from` to `to`.
+     *
+     * Emits a {TransferSingle} event.
+     *
+     * Requirements:
+     *
+     * - `to` cannot be the zero address.
+     * - `from` must have a balance of tokens of type `id` of at least `amount`.
+     * - If `to` refers to a smart contract, it must implement {IERC1155Receiver-onERC1155Received} and return the
+     * acceptance magic value.
+     */
+    function _safeTransferFrom(
+        address from,
+        address to,
+        uint256 id,
+        uint256 amount,
+        bytes memory data
+    ) internal {
+        require(to != address(0), "ForgeTokenFacet: transfer to the zero address");
+
+        uint256 fromBalance = s._balances[id][from];
+        require(fromBalance >= amount, "ForgeTokenFacet: insufficient balance for transfer");
+        unchecked {
+            s._balances[id][from] = fromBalance - amount;
+        }
+        s._balances[id][to] += amount;
+
+        emit TransferSingle(msg.sender, from, to, id, amount);
```

Comment:
It is important to keep consistent in the contracts whether you use LibMeta.msgSender() or msg.sender.